pandas/tests/reshape/merge/test_merge.py

```diff
@@ -959,6 +960,20 @@ def test_merge_on_index_with_more_values(self, how):
         expected.set_index('a', drop=False, inplace=True)
         assert_frame_equal(result, expected)
 
+    def test_merge_right_index_right(self):
+        # https://github.com/pandas-dev/pandas/issues/24897
+        left = pd.DataFrame({'a': [1, 2, 3], 'key': [0, 1, 1]})
+        right = pd.DataFrame({'b': [1, 2, 3]})
+
```

Comment:
Can others double-check this expected output? This is what 0.23.4 / this PR gives. I'm a bit confused about the duplicate `2` in the index.

```python

In [5]: expected
Out[5]:
     a  key  b
0  1.0    0  1
1  2.0    1  2
2  3.0    1  2
2  NaN    2  3
```

I may have expected something like 

```python
In [11]: pd.merge(left, right.rename_axis('key').reset_index(), on='key', how='right').set_index('key', drop=False)
Out[11]:
       a  key  b
key
0    1.0    0  1
1    2.0    1  2
1    3.0    1  2
2    NaN    2  3
```

where the 1 is duplicated. Or am I missing something?

Reply:
Ah, I was first looking at the 'b' column (where the duplicate 2 is of course normal), but the index :)

I assume this is because with `how='right` and `right_index=True`, we use the index of the right for the index of the final dataframe (so the 'key' column corresponds to the index in the result).

Not sure this is fully correct though. I don't know if we have somewhere a specification of what the resulting index of a merge should be.

Reply:
> so the 'key' column corresponds to the index in the result

and again looking wrongly, because what I describe is your expected result, not the actual ...

Yes, so that certainly looks wrong

Reply:
:)

I may just add a comment that this test should probably fail as well, and open a new issue about it, since this is the behavior on 0.23.4.

Reply:
Yes, a comment would be good

Reply:
Hmm this looks vaguely similar to https://github.com/pandas-dev/pandas/issues/17257